third_party/move/move-compiler/src/expansion/ast.rs

```diff
@@ -548,6 +548,44 @@ pub enum Exp_ {
 }
 pub type Exp = Spanned<Exp_>;
 
+impl Exp_ {
+    /// Get all names from an expression
+    /// only recursively check on expression types supported in constant definition.
+    pub fn get_names_for_const_exp(&self) -> HashSet<Name> {
+        let mut names = HashSet::new();
+        let mut add_names = |v: &Exp| {
+            let set = v.value.get_names_for_const_exp();
+            for n in set.iter() {
+                names.insert(*n);
+            }
+        };
+        match self {
+            Self::Name(access, _) => {
+                names.insert(*access.value.get_name());
+            },
+            Self::Call(_, _, _, exp_vec) | Self::Vector(_, _, exp_vec) => {
+                let _ = exp_vec.value.iter().map(&mut add_names);
+            },
+            Self::UnaryExp(_, exp) => {
+                add_names(exp);
+            },
+            Self::BinopExp(exp1, _, exp2) => {
+                add_names(exp1);
+                add_names(exp2);
+            },
+            Self::Block(seq) => {
+                for s in seq.iter() {
+                    if let SequenceItem_::Seq(exp) = &s.value {
+                        add_names(exp);
+                    }
+                }
+            },
+            _ => {},
```

Comment:
Assuming this is only called in the context of const expressions, would it be better to have an internal error/ICE here?

Reply:
If you decide to go the route of adding an error, then the `Block` case has some silently skipped conditions too.

Reply:
I moved the function internally into `analyze_constant` to make sure it will not be called by other functions. Then, all other cases will be handled by `is_valid_for_constant`.